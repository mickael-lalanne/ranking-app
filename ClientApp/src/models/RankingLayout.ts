import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
