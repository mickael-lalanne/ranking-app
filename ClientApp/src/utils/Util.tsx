import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
