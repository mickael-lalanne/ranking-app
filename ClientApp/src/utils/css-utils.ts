export const ELEMENT_SIZE: string = '75px';
