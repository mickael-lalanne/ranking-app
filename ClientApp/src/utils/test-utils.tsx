import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
