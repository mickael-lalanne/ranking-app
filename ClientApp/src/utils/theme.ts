import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
