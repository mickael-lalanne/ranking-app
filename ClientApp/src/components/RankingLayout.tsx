import React, { useEffect, useState } from 'react';
import { css } from '@emotion/css';
import AppButton from './shared/AppButton';
import AppTitle from './shared/AppTitle';
import { useTheme } from '@mui/material/styles';
import { Button } from '@mui/material';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import { DELETE_CONFIRM_TEMPLATE_CONTENT, DELETE_CONFIRM_TEMPLATE_TITLE, DELETE_CONFIRM_TIERLIST_TITLE, ERankingLayoutMode, ERankingLayoutType, RankingLayoutProps, RankingType } from '../models/RankingLayout';
import { getTemplates } from '../services/TemplateServices';
import { useAppDispatch, useAppSelector } from '../app/hooks';
import { getTierlists } from '../services/TierlistServices';
import { useLocation } from 'react-router-dom';
import { useAuth } from "@clerk/clerk-react";
import { updateLoading, updateUser } from '../store/ApplicationStore';
import ConfirmDialog from './shared/ConfirmDialog';
import { Template } from '../models/Template';

const RankingLayout = (
    {
        viewerTitle, viewerSubtitle, viewerBtnText,
        builderTitle, builderSubtitle, builderBtnText,
        editorTitle, editorSubtitle, editorBtnText,
        type,
        ViewerComponent,
        EditorComponent,
        createFunction,
        updateFunction,
        deleteFunction
    } : RankingLayoutProps
) => {
    const [rankingLayoutMode, setRankingLayoutMode] = useState<ERankingLayoutMode>(ERankingLayoutMode.Viewer);
    const [headerTitle, setHeaderTitle] = useState<string>(viewerTitle);
    const [headerSubtitle, setHeaderSubtitle] = useState<string>(viewerSubtitle);
    const [headerButtonText, setHeaderButtonText] = useState<string>(viewerBtnText);
    const [headerButtonColor, setHeaderButtonColor] = useState<string>();
    const [headerButtonDisable, setHeaderButtonDisable] = useState<boolean>(false);
    const [itemToEdit, setItemToEdit] = useState<RankingType>();
    const [showConfirmation, setShowConfirmation] = useState<boolean>(false);
    const [confirmTitle, setConfirmTitle] = useState<string>('');
    const [confirmContent, setConfirmContent] = useState<string>('');

    const dispatch = useAppDispatch();
    const location = useLocation();
    const { userId } = useAuth();

    const templates: Template[] = useAppSelector(state => state.templates.templates);

    // Called when the component is initialized
    useEffect(() => {
        // First, get user info and store it into the store
        dispatch(updateUser({ id: userId}));

        // Then, get all user templates from the database
        const fetchTemplates: () => Promise<void> = async () => await getTemplates(dispatch, userId!);
        fetchTemplates()
            .catch(err => {
                // TODO: handle errors
            });
    }, []);

    // Called when the user come to the Tierlists page
    useEffect(() => {
        if (type === ERankingLayoutType.Tierlist) {
            // Get all user tierlists from the database
            const fetchTierlists: () => Promise<void> = async () => await getTierlists(dispatch);
            fetchTierlists()
                .catch(err => {
                    // TODO: handle errors
                });
        }
    }, [ViewerComponent]);

    // Called when the templates value has changed
    useEffect(() => {
        if (type === ERankingLayoutType.Tierlist) {
            // Dsiable the "Add tierlist" button if there is no templates"
            setHeaderButtonDisable(templates.length === 0);
        } else {
            setHeaderButtonDisable(false);
        }
    }, [templates, type]);

    // Called when the layout mode has changed (for ex: from Viewer to Builder)
    useEffect(() => {
        // Reset the itemToEdit property when leaving the builder / editor mode
        if (rankingLayoutMode === ERankingLayoutMode.Viewer) {
            setItemToEdit(undefined);
        }
    }, [rankingLayoutMode]);

    // Called when the url has changed
    // For example, user goes from Templates page to Tierlists page
    useEffect(() => {
        // Update header and buttons values
       _switchMode(ERankingLayoutMode.Viewer);
    }, [location]);

    const theme = useTheme();
    const loading: boolean = useAppSelector(state => state.application.loading);

    /**
     * Change the mode and the header texts depending on the mode parameter
     * @param {ERankingLayoutMode} mode the new mode to display
     */
    const _switchMode = (mode: ERankingLayoutMode) => {
        switch (mode) {
            case ERankingLayoutMode.Builder:
                setHeaderTitle(builderTitle);
                setHeaderSubtitle(builderSubtitle);
                setHeaderButtonText(builderBtnText);
                setHeaderButtonColor('white');
                break;

            case ERankingLayoutMode.Editor:
                setHeaderTitle(editorTitle);
                setHeaderSubtitle(editorSubtitle);
                setHeaderButtonText(editorBtnText);
                setHeaderButtonColor('white');
                break;
        
            case ERankingLayoutMode.Viewer:
                setHeaderTitle(viewerTitle);
                setHeaderSubtitle(viewerSubtitle);
                setHeaderButtonText(viewerBtnText);
                setHeaderButtonColor(theme.defaultRankingTheme.primary);
                break;
        }
        setRankingLayoutMode(mode);
    }

    /**
     * Called when the header button has been clicked
     * Change the layout mode
     */
    const onHeaderButtonClick = (): void => {
        switch (rankingLayoutMode) {
            case ERankingLayoutMode.Viewer:
                _switchMode(ERankingLayoutMode.Builder);
                break;
        
            case ERankingLayoutMode.Builder:
            case ERankingLayoutMode.Editor:
                _switchMode(ERankingLayoutMode.Viewer);
                break;
        }
    };

    /**
     * Used in the Viewer when a template or tierlist preview has been clicked
     * Show the Editor view to update the item
     * @param {RankingType} itemToEdit the template or tierlist to edit
     */
    const editHandler = (itemToEdit: RankingType): void => {
        setItemToEdit(itemToEdit);
        _switchMode(ERankingLayoutMode.Editor);
    };

    /**
     * Used in the Template Editor when a template needs to be created or updated
     * @param {Template} newItem template to create or update depending on the mode
     */
    const saveHandler = async (newItem: RankingType): Promise<void> => {
        if (rankingLayoutMode === ERankingLayoutMode.Builder) {
            await createFunction(newItem, dispatch);
        }

        if (rankingLayoutMode === ERankingLayoutMode.Editor) {
            await updateFunction(newItem, dispatch);
        }

        _switchMode(ERankingLayoutMode.Viewer);

        // Hide loading indicator
        dispatch(updateLoading(false));
    };

    /**
     * Called when the delete button has been clicked
     * Display a confirmation popup before deleting the item in base
     */
    const onDeleteItemButtonClick = async (): Promise<void> => {
        if (itemToEdit && itemToEdit.id) {
            setConfirmTitle(type === ERankingLayoutType.Template
                ? DELETE_CONFIRM_TEMPLATE_TITLE
                : DELETE_CONFIRM_TIERLIST_TITLE
            );
            setConfirmContent(type === ERankingLayoutType.Template
                ? DELETE_CONFIRM_TEMPLATE_CONTENT
                : ''
            );
            setShowConfirmation(true);
        } else {
            _switchMode(ERankingLayoutMode.Viewer);
        }
    };

    /**
     * Called when the confirm button of the dialog has been clicked
     * Call the server to delete the template or the tierlist in the database
     */
    const deleteConfirmHandler = async (): Promise<void> => {
        if (itemToEdit && itemToEdit.id) {
            // Show loading indicator
            dispatch(updateLoading(true));

            setShowConfirmation(false);

            await deleteFunction(itemToEdit, dispatch);

            // Hide loading indicator
            dispatch(updateLoading(false));
        }
        _switchMode(ERankingLayoutMode.Viewer);
    };

    /**
     * Show Viewer or Builder depending on the current mode
     * @returns {JSX.Element} the Viewer or the Builder React component
     */
    const showRightMode = (): JSX.Element => {
        switch (rankingLayoutMode) {
            case ERankingLayoutMode.Builder:
            case ERankingLayoutMode.Editor:
                return <EditorComponent
                    saveHandler={saveHandler}
                    itemToEdit={itemToEdit}
                    mode={rankingLayoutMode}
                />;

            case ERankingLayoutMode.Viewer:
            default:
                return <ViewerComponent editHandler={editHandler} />
        }
    };

    /**
     * @returns {JSX.Element} a delete button only if we are in edit mode
     */
    const DeleteItemButton = (): JSX.Element => {
        if (rankingLayoutMode === ERankingLayoutMode.Editor) {
            return (
                <Button onClick={onDeleteItemButtonClick} className={delete_btn_style} disabled={loading}>
                    <DeleteForeverIcon style={{ height: '32px', width: '32px' }} />
                </Button>
            );
        }
        return <></>;
    };

    return(<>
        <div className={header_style}>
            <AppTitle title={headerTitle} subtitle={headerSubtitle} />
            <div className="app_spacer"></div>
            {DeleteItemButton()}
            <AppButton
                text={headerButtonText}
                onClickHandler={onHeaderButtonClick}
                color={headerButtonColor}
                disabled={loading || headerButtonDisable}
            />
        </div>

        {showRightMode()}

        <ConfirmDialog
            open={showConfirmation}
            title={confirmTitle}
            content={confirmContent}
            confirmHandler={deleteConfirmHandler}
            cancelHandler={() => setShowConfirmation(false)}
            textValidation={type === ERankingLayoutType.Template ? itemToEdit?.name : undefined}
        />
    </>);
};

export default RankingLayout;

/**
 * CSS STYLES
 */
const header_style = css({
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center'
});

const delete_btn_style = css({
    minWidth: '50px !important',
    height: '50px',
    margin: '0 15px !important'
});
