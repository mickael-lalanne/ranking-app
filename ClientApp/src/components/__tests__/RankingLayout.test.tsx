import * as React from 'react';
import RankingLayout from '../RankingLayout';
import { renderWithProviders } from '../../utils/test-utils';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/react';
import {
    RankingLayoutProps,
    TEMPLATE_LAYOUT_PROPS,
    TIERLIST_LAYOUT_PROPS,
} from '../../models/RankingLayout';
import { act } from 'react-dom/test-utils';

// Mock the server calls from the template service when saving a template
const templateServicePath = '../../services/TemplateServices';
jest.mock(templateServicePath, () => ({
    ...jest.requireActual(templateServicePath),
    getTemplates: () =>
        new Promise<[]>((resolve) => {
            resolve([]);
        }),
}));
const tierlistServicePath = '../../services/TierlistServices';
jest.mock(tierlistServicePath, () => ({
    ...jest.requireActual(tierlistServicePath),
    getTierlists: () =>
        new Promise<[]>((resolve) => {
            resolve([]);
        }),
}));

const RankingLayoutTest = (props: RankingLayoutProps) => {
    return (
        <MemoryRouter>
            <RankingLayout
                viewerTitle={props.viewerTitle}
                viewerSubtitle={props.viewerSubtitle}
                viewerBtnText={props.viewerBtnText}
                builderTitle={props.builderTitle}
                builderSubtitle={props.builderSubtitle}
                builderBtnText={props.builderBtnText}
                editorTitle={props.editorTitle}
                editorSubtitle={props.editorSubtitle}
                editorBtnText={props.editorBtnText}
                type={props.type}
                ViewerComponent={props.ViewerComponent}
                EditorComponent={props.EditorComponent}
                createFunction={props.createFunction}
                updateFunction={props.updateFunction}
                deleteFunction={props.deleteFunction}
            />
        </MemoryRouter>
    );
};

describe('Screenshot', () => {
    it('takes template screenshot', async () => {
        let component;
        await act(async () => {
            component = renderWithProviders(
                RankingLayoutTest(TEMPLATE_LAYOUT_PROPS)
            );
        });

        expect(component!.baseElement).toMatchSnapshot('template-layout');
    });

    it('takes tierlist screenshot', async () => {
        let component;
        await act(async () => {
            component = renderWithProviders(
                RankingLayoutTest(TIERLIST_LAYOUT_PROPS)
            );
        });

        expect(component!.baseElement).toMatchSnapshot('tierlist-layout');
    });
});

describe('Common', () => {
    it('switchs from viewer to builder mode', async () => {
        await act(async () => {
            renderWithProviders(RankingLayoutTest(TEMPLATE_LAYOUT_PROPS));
        });

        // Click on the header button
        await userEvent.click(
            screen.getByText(TEMPLATE_LAYOUT_PROPS.viewerBtnText)
        );

        // Now, we should be in the editor view
        expect(
            screen.getByText(TEMPLATE_LAYOUT_PROPS.builderBtnText)
        ).toBeDefined();
    });
});
