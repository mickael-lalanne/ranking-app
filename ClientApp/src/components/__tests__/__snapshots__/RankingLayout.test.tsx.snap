// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes template screenshot: template-layout 1`] = `
<body>
  <div>
    <div
      class="css-69i1ev"
    >
      <div
        class="css-1anxslm"
      >
        <h1>
          My templates
          <span>
            Create, Edit or Delete
          </span>
        </h1>
      </div>
      <div
        class="app_spacer"
      />
      <div
        class="css-70qvj9"
        data-cy="header-buttons"
      >
        <button
          class="
                css-1t5c0s4
                
            "
          data-cy="app-button"
          role="button"
        >
          <span
            class="css-1uxuv1w add_btn_shadow"
          />
          <span
            class="css-e5grz5"
          />
          <span
            class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
            style="color: white;"
          >
            Add template
          </span>
        </button>
      </div>
    </div>
    <div
      class="css-afv6yl"
      data-cy="info-box"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1695c0c css-i4bv87-MuiSvgIcon-root"
        data-testid="TipsAndUpdatesIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 20h4c0 1.1-.9 2-2 2s-2-.9-2-2zm-2-1h8v-2H5v2zm11.5-9.5c0 3.82-2.66 5.86-3.77 6.5H5.27c-1.11-.64-3.77-2.68-3.77-6.5C1.5 5.36 4.86 2 9 2s7.5 3.36 7.5 7.5zm4.87-2.13L20 8l1.37.63L22 10l.63-1.37L24 8l-1.37-.63L22 6l-.63 1.37zM19 6l.94-2.06L22 3l-2.06-.94L19 0l-.94 2.06L16 3l2.06.94L19 6z"
        />
      </svg>
      <span>
        It seems you don't have any template yet. 
        <br />
        Click on the button above to start your creation !
      </span>
    </div>
    <div
      class="css-1h8dwix"
      data-testid="templates-viewer"
    />
  </div>
</body>
`;

exports[`Screenshot takes tierlist screenshot: tierlist-layout 1`] = `
<body>
  <div>
    <div
      class="css-69i1ev"
    >
      <div
        class="css-1anxslm"
      >
        <h1>
          My tierlists
          <span>
            Create, Edit or Delete
          </span>
        </h1>
      </div>
      <div
        class="app_spacer"
      />
      <div
        class="css-70qvj9"
        data-cy="header-buttons"
      >
        <button
          class="
                css-1t5c0s4
                css-av1nc5
            "
          data-cy="app-button"
          disabled=""
          role="button"
        >
          <span
            class="css-1uxuv1w add_btn_shadow"
          />
          <span
            class="css-e5grz5"
          />
          <span
            class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
            style="color: white;"
          >
            Add tierlist
          </span>
        </button>
      </div>
    </div>
    <div>
      <div
        class="css-1yd03il"
        data-testid="tierlists-viewer"
      >
        <div
          class="css-afv6yl"
          data-cy="info-box"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1695c0c css-i4bv87-MuiSvgIcon-root"
            data-testid="TipsAndUpdatesIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 20h4c0 1.1-.9 2-2 2s-2-.9-2-2zm-2-1h8v-2H5v2zm11.5-9.5c0 3.82-2.66 5.86-3.77 6.5H5.27c-1.11-.64-3.77-2.68-3.77-6.5C1.5 5.36 4.86 2 9 2s7.5 3.36 7.5 7.5zm4.87-2.13L20 8l1.37.63L22 10l.63-1.37L24 8l-1.37-.63L22 6l-.63 1.37zM19 6l.94-2.06L22 3l-2.06-.94L19 0l-.94 2.06L16 3l2.06.94L19 6z"
            />
          </svg>
          <span>
            It seems you don't have any template yet. 
            <br />
            Start by going to the templates page !
          </span>
        </div>
      </div>
      <div />
    </div>
  </div>
</body>
`;
